api4/resolver_channel_member_test.go

```diff
@@ -0,0 +1,355 @@
+	myTeam := th.CreateTeam()
+	ch1 := th.CreateChannelWithClientAndTeam(th.Client, model.ChannelTypeOpen, myTeam.Id)
+	ch2 := th.CreateChannelWithClientAndTeam(th.Client, model.ChannelTypePrivate, myTeam.Id)
+	th.LinkUserToTeam(th.BasicUser, myTeam)
+	th.App.AddUserToChannel(th.BasicUser, ch1, false)
+	th.App.AddUserToChannel(th.BasicUser, ch2, false)
+
+	// Creating some msgcount
+	th.CreateMessagePostWithClient(th.Client, th.BasicChannel, "basic post")
+	th.CreateMessagePostWithClient(th.Client, ch1, "ch1 post")
+
+	var q struct {
+		ChannelMembers []struct {
+			Channel struct {
+				ID          string            `json:"id"`
+				CreateAt    float64           `json:"createAt"`
+				UpdateAt    float64           `json:"updateAt"`
+				Type        model.ChannelType `json:"type"`
+				DisplayName string            `json:"displayName"`
+				Name        string            `json:"name"`
+				Header      string            `json:"header"`
+				Purpose     string            `json:"purpose"`
+				Team        struct {
+					ID string `json:"id"`
+				} `json:"team"`
+			} `json:"channel"`
+			User struct {
+				ID        string `json:"id"`
+				Username  string `json:"username"`
+				Email     string `json:"email"`
+				FirstName string `json:"firstName"`
+				LastName  string `json:"lastName"`
+				NickName  string `json:"nickname"`
+			} `json:"user"`
+			Roles []struct {
+				ID            string   `json:"id"`
+				Name          string   `json:"Name"`
+				Permissions   []string `json:"permissions"`
+				SchemeManaged bool     `json:"schemeManaged"`
+				BuiltIn       bool     `json:"builtIn"`
+			} `json:"roles"`
+			LastViewedAt     float64         `json:"lastViewedAt"`
+			LastUpdateAt     float64         `json:"lastUpdateAt"`
+			MsgCount         float64         `json:"msgCount"`
+			MentionCount     float64         `json:"mentionCount"`
+			MentionCountRoot float64         `json:"mentionCountRoot"`
+			NotifyProps      model.StringMap `json:"notifyProps"`
+			SchemeGuest      bool            `json:"schemeGuest"`
+			SchemeUser       bool            `json:"schemeUser"`
+			SchemeAdmin      bool            `json:"schemeAdmin"`
+			Cursor           string          `json:"cursor"`
+		} `json:"channelMembers"`
+	}
+
+	t.Run("all", func(t *testing.T) {
+		input := graphQLInput{
+			OperationName: "channelMembers",
+			Query: `
+	query channelMembers {
+	  channelMembers(userId: "me") {
+	  	channel {
+		  	id
+		  	createAt
+		  	updateAt
+		  	type
+		    displayName
+		    name
+		    header
+		    team {
+		    	id
+		    }
+	  	}
+	  	user {
+	  		id
+	  		username
+	  		email
+	  	}
+	  	msgCount
+	  	mentionCount
+	  	mentionCountRoot
+	  	schemeGuest
+	  	schemeUser
+	  	schemeAdmin
+	  	cursor
+	  }
+	}
+	`,
+		}
+
+		resp, err := th.MakeGraphQLRequest(&input)
+		require.NoError(t, err)
+		require.Len(t, resp.Errors, 0)
+		require.NoError(t, json.Unmarshal(resp.Data, &q))
+		assert.Len(t, q.ChannelMembers, 9)
```

Comment:
Instead of using magic numbers here, would you think it'd be more robust to get these numbers upfront? Any change on the testlib regarding to users may break these tests right?

Reply:
Hmm .. how do you mean "get these numbers upfront"? These are actually the same API tests for REST. See `TestGetChannelsForUser` in api4/channel_test.go.

This basically comes from the 2 initial channels (town-square+off-topic) and then the additional teams and channels we create in L23-28. Unless we change the initial set of channels, these shouldn't break. And even if we do change the initial channels, I'd expect some tests to break.

Reply:
It was actually meant that maybe we should validating these against the REST API. The flow would be like: getChannelCount (1) -> do the test (2) -> compare with step 1 (3). But if that is a common pattern in our codebase, you can ignore my comment 😅

Reply:
Ahh I see. Well, then the question is who guarantees the REST API is correct? Because the exact same test is for the REST API as well :stuck_out_tongue: If you are relying on the REST API being correct, then that shouldn't make much difference because both use the same test.

Reply:
Ehe, we should rely on REST API at some point 😅 But it's not a big deal anyway.